apps/mobile-wallet/src/persistent-storage/wallet.ts

```diff
@@ -56,6 +57,118 @@ const MNEMONIC_V2 = 'wallet-mnemonic-v2'
+    mnemonicV2Exists = await storedMnemonicV2Exists()
+  } catch {
+    mnemonicV2Exists = false
+  }
+
+  if (mnemonicV2Exists) {
+    if (walletMetadata) {
+      // If we have both mnemonic and metadata available, then all good
+      return true
+    } else {
+      // If we have mnemonic but missing metadata, we try to recreate them with sane defaults
+      try {
+        await generateAndStoreWalletMetadata('My wallet', false)
+      } finally {
+        walletMetadata = await getWalletMetadata(false)
+      }
+
+      if (walletMetadata) {
+        showToast({
+          text1: i18n.t('App data were reset'),
+          text2: i18n.t(
+            'Due to an unexpected error some of your app data were reset. You might want to scan for active addresses in the settings.'
+          ),
+          type: 'info',
+          autoHide: false
+        })
+        sendAnalytics({ event: 'Recreated missing wallet metadata for existing wallet' })
+
+        return true
+      } else {
+        showToast({
+          text1: i18n.t('Could not unlock app'),
+          text2: i18n.t('Wallet metadata not found'),
+          type: 'error',
+          autoHide: false
+        })
+        sendAnalytics({ type: 'error', message: 'Could not recreate missing wallet metadata for existing wallet' })
+
+        return false
+      }
+    }
+  } else {
+    let deprecatedWalletExists
+
+    try {
+      deprecatedWalletExists = !!(await getSecurelyWithReportableError(PIN_WALLET_STORAGE_KEY, true, ''))
+    } catch {
+      deprecatedWalletExists = false
+    }
+
+    if (deprecatedWalletExists) {
+      if (walletMetadata) {
+        // If we have no mnemonic, but we have a deprecated one with metadata, all good, the pin/bio flow will migrate
+        return true
+      } else {
+        // If we only have a deprecated mnemonic but no metadata we recreate deprecated metadata with sane defaults and
+        // the migration flow will migrate both mnemonic and metadata
+        try {
+          await storeWalletMetadataDeprecated({
+            id: nanoid(),
+            name: 'My wallet',
+            isMnemonicBackedUp: false,
+            addresses: [
+              {
+                index: 0,
+                isDefault: true,
+                color: getRandomLabelColor()
+              }
+            ],
+            contacts: []
+          })
+
+          return true
+        } catch {
+          return false
+        }
+      }
+    } else {
+      if (!walletMetadata) {
+        // If we have no mnemonic, no deprecated mnemonic, and no metadata, all good, fresh install
+        return true
+      } else {
+        // If we have metadata but no mnemonic and no deprecated mnemonic, we should delete the metadata
+        try {
+          await deleteWithReportableError(WALLET_METADATA_STORAGE_KEY)
+        } finally {
+          walletMetadata = await getWalletMetadata(false)
+        }
+
+        if (!walletMetadata) {
+          return true
+        } else {
```

Comment:
Nit: I like how you commented the flow above. One more comment here would be nice! 
"This should never happen. Could not delete metadata for some reason and we don't have any mnemonic, we can't unlock"

Reply:
done